In [69]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

In [13]:
df = pd.read_pickle("./data.pkl")
df.head()

1267     1268     1270     1271     1272     1273     1274     \
Gis1-24        0.0      0.0      0.0      0.0      0.0      0.0      0.0   
Bwi1-74        0.0      0.0      0.0      0.0      0.0      0.0      0.0   
Tai_R2-30      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
Sob1-31        0.0      0.0      0.0      0.0      0.0      0.0      0.0   
LCA-3-12       0.0      0.0      0.0      0.0      0.0      0.0      0.0   

           1451     1452     1453     ...  2690716  2690717  2690718  2690719  \
Gis1-24        0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   
Bwi1-74        0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   
Tai_R2-30      0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   
Sob1-31        0.0      0.0      0.0  ...      0.0      0.0      0.0      0.0   
LCA-3-12       NaN      NaN      NaN  ...      0.0      0.0      0.0      0.0   

           2690720  2690721  2690722  2690723  2690724  2690725  
Gis1-24        0.0      0.0      2.0      0.0      0.0      0.0  
Bwi1-74        0.0      0.0      2.0      0.0      0.0      0.0  
Tai_R2-30      0.0      0.0      2.0      0.0      0.0      0.0  
Sob1-31        0.0      0.0      2.0      0.0      0.0      0.0  
LCA-3-12       0.0      0.0      2.0      0.0      0.0      0.0  

[5 rows x 993247 columns]

In [20]:
df.stack().value_counts().to_dict()

{0.0: 240162970, 2.0: 17846959, 1.0: 915310}

In [21]:
df.shape

(280, 993247)

In [31]:
X = df.reset_index(drop = True)
X = X.fillna(X.mean())

NameError: name 're' is not defined

In [34]:
import re

y = []

for site in df.index.values:
    y.append(re.sub('[^a-zA-Z]+', '', site))

y

['Gis',
 'Bwi',
 'TaiR',
 'Sob',
 'LCA',
 'Baf',
 'Gou',
 'N',
 'Din',
 'Djo',
 'El',
 'Gas',
 'Baf',
 'Baf',
 'Bil',
 'Fjn',
 'Nim',
 'Fjn',
 'San',
 'Boe',
 'Gbo',
 'Gas',
 'Kay',
 'GB',
 'Kor',
 'Kay',
 'Gas',
 'Boe',
 'Lop',
 'Con',
 'Gep',
 'Kab',
 'Bud',
 'Gbo',
 'Sob',
 'Gis',
 'Nim',
 'Nim',
 'Bud',
 'San',
 'CNPW',
 'CNPE',
 'Mtc',
 'Gbo',
 'Gis',
 'Baf',
 'Djo',
 'Baf',
 'Uga',
 'Gis',
 'Fjn',
 'Gis',
 'Nim',
 'Baf',
 'Bat',
 'N',
 'Con',
 'Bud',
 'MTC',
 'Bwi',
 'Cam',
 'Kor',
 'Nim',
 'Fouta',
 'Mtc',
 'TaiR',
 'Kay',
 'N',
 'NNP',
 'Cnp',
 'Gep',
 'Cnp',
 'Rt',
 'Mbe',
 'Lib',
 'Djo',
 'Gis',
 'N',
 'Din',
 'Bil',
 'Gas',
 'Uga',
 'Gis',
 'Mbe',
 'Gis',
 'Dja',
 'Gis',
 'Baf',
 'Fouta',
 'Fouta',
 'GB',
 'San',
 'Onp',
 'Lib',
 'NNP',
 'Gis',
 'Gou',
 'Mtc',
 'Gbo',
 'Mtc',
 'Uga',
 'CNPW',
 'Gep',
 'Bud',
 'TaiR',
 'Djo',
 'Gis',
 'Gas',
 'Mtc',
 'Con',
 'Boe',
 'Uga',
 'Uga',
 'Boe',
 'N',
 'Gis',
 'Fjn',
 'Gou',
 'TaiE',
 'Kay',
 'Boe',
 'Uga',
 'Boe',
 'N',
 'Sob',
 'K

In [42]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
print(X_train.shape)
print(X_test.shape)
print(len(y_train))
print(len(y_test))

(224, 993247)
(56, 993247)
224
56


# Multi-Layer Perceptron

In [73]:
# Train
mlp_clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(5, 2), random_state=1)
mlp_clf.fit(X_train, y_train)

MLPClassifier(alpha=1e-05, hidden_layer_sizes=(5, 2), random_state=1,
              solver='lbfgs')

In [74]:
# Predict + get accuracy
mlp_y_pred = mlp_clf.predict(X_test)
print(accuracy_score(y_test, mlp_y_pred))

0.017857142857142856


The results provided by using a Multi-Layer Perceptron are notoriously bad, resulting in an accuracy of around 0.01%.

# Random Forest:

In [79]:
# Train
rf_clf = RandomForestClassifier()
rf_clf.fit(X_train, y_train)

RandomForestClassifier()

In [72]:
# Predict + get accuracy
rf_y_pred = rf_clf.predict(X_test)
print(accuracy_score(y_test, rf_y_pred))

0.6071428571428571


The Random Forest classifier seems promising. It yields an accuracy of around 60.7% using bootstrapping, and around 57.1% without it.

In [84]:
len(y)
1 / 280

0.0035714285714285713

The chance accuracy is ~0.003